tool/src/org/antlr/v4/automata/optimization/RedundantEpsilonRemover.java

```diff
@@ -0,0 +1,180 @@
+package org.antlr.v4.automata.optimization;
+
+import org.antlr.v4.runtime.atn.*;
+
+import java.util.*;
+
+public class RedundantEpsilonRemover {
+	public final ATN atn;
+
+	private final ATNOptimizerHelper helper;
+
+	private RedundantEpsilonRemover(ATN atn, ATNOptimizerHelper helper) {
+		this.atn = atn;
+		this.helper = helper;
+	}
+
+	public static int optimize(ATNOptimizerHelper helper) {
+		List<ATNState> removedStates = helper.getRemovedStates();
+		int numberOfRemovedStates = removedStates.size();
+		new RedundantEpsilonRemover(helper.atn, helper).traverseATN();
+		return removedStates.size() - numberOfRemovedStates;
+	}
+
+	private void traverseATN() {
+		Set<ATNState> visitedStates = new HashSet<>();
+		for (RuleStartState start : atn.ruleToStartState) {
+			traverseState(start, visitedStates);
+		}
+	}
+
+	private void traverseState(ATNState state, Set<ATNState> visitedStates) {
+		if (!visitedStates.add(state)) {
+			return;
+		}
+
+		optimizeState(state);
+
+		for (int i = 0; i < state.getNumberOfTransitions(); i++) {
+			Transition transition = state.transition(i);
+			traverseState(transition.target, visitedStates);
+			if (transition instanceof RuleTransition) {
+				traverseState(((RuleTransition) transition).followState, visitedStates);
+			}
+		}
+	}
+
+	private void optimizeState(ATNState state) {
+		if (state instanceof BasicState && state.getNumberOfTransitions() == 1) {
+			Transition outTransition = state.transition(0);
+			Boolean removeInTransition = null;
+
+			List<InTransition> inTransitions = helper.getInTransitions(state);
+			InTransition inTransition0 = inTransitions.get(0);
+
+			if (inTransitions.size() == 1) {
+				/*
+				  Input:
+				      ε     t
+				  I1 --> S --> O1
+
+				  Output:
+				      t
+				  I1 --> O1
+				*/
```

Comment:
Inline comments usually use single line comment style `//`. Only if you want to comment out something temporarily the multiline comment style is used. At least that's what I use.